Sources/System/FilePath/FilePathSyntax.swift

```diff
@@ -0,0 +1,353 @@
+  ///
+  /// Examples:
+  /// * Unix:
+  ///   * `~/bar`
+  ///   * `tmp/foo.txt`
+  /// * Windows:
+  ///   * `bar\baz`
+  ///   * `C:Users\`
+  ///   * `\Users`
+  public var isRelative: Bool { !isAbsolute }
+
+  /// A textual representation of the file path using the generic separator (`/`).
+  ///
+  /// On Unix, this is equivalent to `description`
+  ///
+  /// TODO: Should `dump()` use this? Also, should dump escape invalid Unicode?
+  public var genericDescription: String {
+    fatalError()
+  }
+
+  /// Returns whether `self` is a prefix of `other`
+  public func starts(with other: FilePath) -> Bool {
+    fatalError()
+  }
+
+  /// Returns whether `self` is a suffix of `other`
+  public func ends(with other: FilePath) -> Bool {
+    fatalError()
+  }
+
+  /// Whether this path is empty
+  public var isEmpty: Bool { self.storage == [0] }
+}
+
+
+// Decompose a path into its parts
+//
+//   path = root? relative-path?
+//   root = root-name? root-directory?
+//   root-directory = separator
+//   root-name (Windows only) = disk | UNC | device
+extension FilePath {
+  /// Returns the root directory of a path if there is one, otherwise `nil`
+  ///
+  /// On Unix, this will return the leading `/` if the path is absolute
+  /// and `nil` if the path is relative.
+  ///
+  /// On Windows, this will return the path prefix up to and including the
+  /// host and share for UNC paths or the volume for device paths. This will
+  /// not return any subsequence separator for UNC or device paths.
+  ///
+  /// On Windows, for traditional DOS paths, this will return
+  /// the path prefix up to and including a root directory or
+  /// a supplied drive or volume. Otherwise, if the path is relative to
+  /// both the current directory and current driver, returns `nil`
+  ///
+  /// Examples:
+  /// * Unix:
+  ///   * `/foo/bar => /`
+  ///   * `foo/bar => nil`
+  /// * Windows:
+  ///   * `C:\foo\bar => C:\`
+  ///   * `C:foo\bar => C:`
+  ///   * `\foo\bar => \ `
+  ///   * `foo\bar => nil`
+  ///   * `\\?\UNC\server\share\file => \\?\UNC\server\share`
+  ///
+  /// TODO: Document setter. Better to document for a path what is root and what is relative.
+  ///
+  /// REVIEW NOTE: C# calls this the root. C++ implies they pull in the trailing separator (I
+  /// haven't run tests). Rust differs in that has_root is just looking for the `\` somewhere.
+  ///
+  /// REVIEW NOTE: Going with UNC or device paths not including trailing separator,
+  /// which is consistent with C# IIUC. I need to perform the experiement to verify this.
+  ///
+  /// REVIEW NOTE: I'm going with component as return type, which implies the more common
+  /// use would be to read or inspect the root, rather than forming new paths from the same root.
+  /// For that use case, you can do `FilePath(otherPath.root)`, or use
+  /// the `relativePath`'s setter, or use the `/` operator (if we go that route).
+  public var root: FilePath.Component? {
```

Comment:
This makes a lot of sense to be based on my understanding of Windows paths.

Reply:
If we _did_ have `AbsolutePath` and `RelativePath`, this is the kind of property that would make sense to have as non-optional for `AbsolutePath` and that just wouldn't be on the `RelativePath` type at all.  Makes things a bit cleaner for the client when they know they have an absolute path, for example (because that's the contract and it has been asserted) and then they can just access the root without having to unwrap.

Reply:
Relative paths on Windows can have roots: `C:` and `\`.

Reply:
Ah, right, that's true.  I forgot about that, and that certainly complicates things.